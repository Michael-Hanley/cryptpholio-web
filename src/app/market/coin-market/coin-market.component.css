.mat-header-cell.mat-sort-header-sorted {
  color: black;
}
.mat-table {
  overflow: auto;
  min-width: 875px;
}
.search-box {
  width: 200px;
  padding: 8px 24px;
  float: left;
}
.container {
  display: flex;
  flex-direction: column;
  min-width: 350px;
  max-width: 100%;
  max-height: 94vh;
  overflow: auto;
  margin-bottom: 20px;
}
.mat-form-field {
  font-size: 14px;
  width: 100%;
}
.mat-column-icon {
  flex: 0 0 45px;
}
.mat-column-symbol {
  flex: 0 0 90px;
}
.mat-column-btc_price {
  flex: 0 0 115px;
}
.mat-column-usd_price {
  flex: 0 0 115px;
}
.spinner-margin {
  margin: 0 10px;
}
.column {
  cursor: pointer;
}
.icon {
  width: 12px;
  height: 12px;
}
.global_market_cap {
  width: 500px;
  padding: 8px 24px;  
  float: right;
  text-align: right;
}