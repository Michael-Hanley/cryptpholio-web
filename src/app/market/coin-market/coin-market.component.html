<div class="container">
  <div>
    <span class="global_market_cap">Global Market Cap: ${{ globalStats?.total_mcap | number: '1.2-2' }}</span>    
    <span class="search-box">
      <mat-form-field>
        <input type="text" matInput (keyup)="applyFilter($event.target.value)" placeholder="Filter">
      </mat-form-field>
    </span>
  </div>
  <mat-table #table [dataSource]="dataSource" matSort>
    <ng-container matColumnDef="icon" >
        <mat-header-cell *matHeaderCellDef></mat-header-cell>
        <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'> <img src="{{coin.icon}}" width="30" height="30"> </mat-cell>
    </ng-container>
    <ng-container matColumnDef="name">
      <mat-header-cell *matHeaderCellDef> Name </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.name}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="symbol">
      <mat-header-cell *matHeaderCellDef> Symbol </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.symbol}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="price">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Price USD </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>${{coin.price | number:'1.2-2'}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="priceBtc">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Price BTC </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'> <mat-icon class="icon" svgIcon="btc"></mat-icon>{{coin.priceBtc | number:'.2-9'}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="marketCap">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Market Cap USD </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>${{coin.marketCap | number:'1.2-2'}}</mat-cell>
    </ng-container>
    
    <ng-container matColumnDef="priceChange1d">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Percent Change </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.priceChange1d | number}}%</mat-cell>
    </ng-container>
    

    <ng-container matColumnDef="availableSupply">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Available Supply </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.availableSupply}}</mat-cell>
    </ng-container>

    <mat-header-row *matHeaderRowDef="columnsToDisplay"></mat-header-row>
    <mat-row *matRowDef="let coins; columns: columnsToDisplay"></mat-row>
  </mat-table>
  <div fxLayout fxLayoutAlign="center center" >
    <mat-spinner class="spinner" *ngIf="!ready">
    </mat-spinner>
  </div>
  <mat-paginator #paginator
    (page)="updateUrl($event)"
    [pageIndex] = "this.pageIndex"
    [length]="100"
    [pageSize]="this.pageSize"
    [pageSizeOptions]="[5, 10, 25, 100]">
  </mat-paginator>
</div>
