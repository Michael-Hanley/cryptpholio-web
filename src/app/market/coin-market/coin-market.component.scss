.mat-header-cell.mat-sort-header-sorted {
  color: black;
}
.search-box {
  width: 200px;
  padding: 8px 24px;
  float: left;
}
.container {
  display: flex;
  flex-direction: column;
  min-width: 350px;
  max-width: 100%;
  max-height: 94vh;
  margin-bottom: 20px;
}
.spinner-margin {
  margin: 0 10px;
}
.global_market_cap {
  width: 500px;
  padding: 8px 24px;  
  float: right;
  text-align: right;
}