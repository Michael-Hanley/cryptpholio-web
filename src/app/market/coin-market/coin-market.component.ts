import { Component, OnInit, ViewChild, OnDestroy } from '@angular/core';
import { CoinService } from '../../services/coin-service.service';
import { MatPaginator, MatTableDataSource, MatSort } from '@angular/material';
import { Router, ActivatedRoute } from '@angular/router';
import { MatIconRegistry } from '@angular/material';
import { DomSanitizer } from '@angular/platform-browser';

import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

@Component({
  selector: 'app-coin-market',
  templateUrl: './coin-market.component.html',
  styleUrls: ['./coin-market.component.css']
})
export class CoinMarketComponent implements OnInit, OnDestroy {
  coins;
  timer;
  columnsToDisplay = [
    'icon',
    'name',
    'symbol',
    'price',
    'priceBtc',
    'marketCap',
    'priceChange1d',
    'availableSupply'
  ];
  imageUrl = 'https://www.cryptocompare.com';
  dataSource = new MatTableDataSource<Element>(this.coins);
  pageIndex;
  pageSize = 10;
  page;
  globalStats: any;
  ready = false;

  private ngUnsubscribe: Subject<any> = new Subject();

  constructor(private coinService: CoinService, private router: Router,
    iconRegistry: MatIconRegistry, sanitizer: DomSanitizer,
    private route: ActivatedRoute) {
    this.getCoins();
    this.getGlobalStats();
    iconRegistry.addSvgIcon(
      'btc',
      sanitizer.bypassSecurityTrustResourceUrl('assets/icons/btc.svg'));

      this.route.queryParams.subscribe(
        params => {
          if (!params.pageSize) {
            this.pageSize = 10;
          } else {
          this.pageIndex = params.pageIndex;
          this.pageSize = params.pageSize;
          }
      });
   }

  @ViewChild(MatPaginator, { static: true }) paginator: MatPaginator;
  @ViewChild(MatSort, { static: true }) sort: MatSort;
  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // MatTableDataSource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }
  coinView(coin) {
    clearTimeout(this.timer);
    this.router.navigate(['/coin'], {queryParams: coin});
  }
  getCoins() {
    this.coinService.getCoins()
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe(coins => {
        if (!coins) { return; }
        this.coins = coins;
        this.dataSource = new MatTableDataSource<Element>(this.coins.coins);
        this.dataSource.paginator = this.paginator;
        this.dataSource.sort = this.sort;
        this.ready = true;
        this.syncCoins();
      });
  }
  getGlobalStats() {
    this.coinService.getGlobalStats()
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe(res => {
        this.globalStats = res[0];
      });
  }
  syncCoins() {
    this.timer = setTimeout(() => this.getCoins(), 60000);
  }
  updateUrl(params) {
    const queryParams = params;
    this.router.navigate(['.'], { queryParams: queryParams });
  }
  ngOnInit() {
    if (this.page !== undefined) {
      this.pageSize = this.pageIndex.pageSize;
      this.pageIndex = this.pageIndex.pageIndex;
    }
  }
  ngOnDestroy(): any {
    this.ngUnsubscribe.next();
    this.ngUnsubscribe.complete();
  }
}
export interface Element {
  name: string;
  symbol: number;
  priceBtc: number;
  price: number;
  marketCap: number;
  priceChange1h: number;
  priceChange1d: number;
  priceChange1w: number;
  icon: string;
  availableSupply: number;
  totalSupply: number;
}
