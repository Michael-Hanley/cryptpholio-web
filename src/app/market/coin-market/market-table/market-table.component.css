.icon {
    width: 12px;
    height: 12px;
}
.mat-table {
    overflow: auto;
    min-width: 875px;
}
.mat-form-field {
    font-size: 14px;
    width: 100%;
}
.mat-column-icon {
    flex: 0 0 45px;
}
.mat-column-symbol {
    flex: 0 0 90px;
}
.mat-column-priceBtc {
    flex: 0 0 115px;
}
.mat-column-price {
    flex: 0 0 95px;
}

.mat-column-marketCap {
    flex: 0 0 155px;
}
.mat-column-priceChange1d {
    flex: 0 0 85px;
    justify-content: center;
}
.column {
    cursor: pointer;
}