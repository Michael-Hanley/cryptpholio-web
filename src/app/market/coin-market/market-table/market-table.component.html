<mat-table #table [dataSource]="dataSource" matSort>
    <ng-container matColumnDef="icon" >
        <mat-header-cell *matHeaderCellDef></mat-header-cell>
        <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'> <img src="{{coin.icon}}" width="30" height="30"> </mat-cell>
    </ng-container>
    <ng-container matColumnDef="name">
      <mat-header-cell *matHeaderCellDef> Name </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.name}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="symbol">
      <mat-header-cell *matHeaderCellDef> Symbol </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.symbol}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="price">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Price USD </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>${{coin.price | number:'1.2-2'}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="priceBtc">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Price BTC </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'> <mat-icon class="icon" svgIcon="btc"></mat-icon>{{coin.priceBtc | number:'.2-9'}}</mat-cell>
    </ng-container>

    <ng-container matColumnDef="marketCap">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Market Cap USD </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>${{coin.marketCap | number:'1.2-2'}}</mat-cell>
    </ng-container>
    
    <ng-container matColumnDef="priceChange1d">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Percent Change 24h</mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.priceChange1d | number}}%</mat-cell>
    </ng-container>
    

    <ng-container matColumnDef="availableSupply">
      <mat-header-cell *matHeaderCellDef mat-sort-header> Available Supply </mat-header-cell>
      <mat-cell class="column" *matCellDef="let coin" (click)='coinView(coin)'>{{coin.availableSupply}}</mat-cell>
    </ng-container>

    <mat-header-row *matHeaderRowDef="columnsToDisplay"></mat-header-row>
    <mat-row *matRowDef="let coins; columns: columnsToDisplay"></mat-row>
  </mat-table>