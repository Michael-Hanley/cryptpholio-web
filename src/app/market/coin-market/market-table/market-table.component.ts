import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { Router } from '@angular/router';

@Component({
  selector: 'app-market-table',
  templateUrl: './market-table.component.html',
  styleUrls: ['./market-table.component.scss']
})
export class MarketTableComponent implements OnInit {
  @Input() dataSource;
  @Input() coins;
  @Input() columnsToDisplay;
  @Output() navigated = new EventEmitter<boolean>();

  constructor(private router: Router) { }

  coinView(coin) {
    this.navigated.emit(true);
    this.router.navigate(['/coin'], {queryParams: {coinId: coin.id}});
  }

  ngOnInit() {
  }

}
