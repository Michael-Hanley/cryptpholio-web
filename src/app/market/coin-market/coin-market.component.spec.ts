import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MockComponent } from 'ng-mocks';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

import { CoinMarketComponent } from './coin-market.component';
import { MarketTableComponent } from '../../market/coin-market/market-table/market-table.component';

import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatPaginatorModule } from '@angular/material/paginator';
import { MatInputModule } from '@angular/material';
import 'hammerjs';

describe('CoinMarketComponent', () => {
  let component: CoinMarketComponent;
  let fixture: ComponentFixture<CoinMarketComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        CoinMarketComponent,
        MockComponent(MarketTableComponent)
      ],
      imports: [
        MatProgressSpinnerModule,
        MatPaginatorModule,
        HttpClientTestingModule,
        RouterTestingModule,
        MatInputModule,
        BrowserAnimationsModule
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CoinMarketComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
