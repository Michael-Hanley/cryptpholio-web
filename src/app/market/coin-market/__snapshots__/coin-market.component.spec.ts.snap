// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinMarketComponent should match snapshot 1`] = `
<app-coin-market
  coinService={[Function CoinService]}
  columnsToDisplay={[Function Array]}
  dataSource={[Function MatTableDataSource]}
  imageUrl={[Function String]}
  ngUnsubscribe={[Function Subject]}
  pageSize={[Function Number]}
  paginator="undefined"
  ready="false"
  route={[Function ActivatedRoute]}
  routeService={[Function RouteService]}
  router={[Function Router]}
  sort="undefined"
>
  <div
    class="container"
  >
    <div>
      <span
        class="global_market_cap"
      >
        Global Market Cap: $
      </span>
      <span
        class="search-box"
      >
        <mat-form-field
          class="mat-form-field ng-tns-c8-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                  id="mat-input-1"
                  matinput=""
                  placeholder="Filter"
                  type="text"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-1"
                    class="mat-form-field-label ng-tns-c8-0 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-1"
                    id="mat-form-field-label-3"
                  >
                    
                    
                    Filter
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-underline ng-tns-c8-0 ng-star-inserted"
            >
              <span
                class="mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              
              <div
                class="mat-form-field-hint-wrapper ng-tns-c8-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
      </span>
    </div>
    <app-market-table />
    <div
      fxlayout=""
      fxlayoutalign="center center"
    >
      
      <mat-spinner
        class="spinner mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation ng-star-inserted"
        mode="indeterminate"
        role="progressbar"
        style="width: 100px; height: 100px;"
      >
        <svg
          focusable="false"
          preserveAspectRatio="xMidYMid meet"
          style="width: 100px; height: 100px;"
          viewBox="0 0 100 100"
        >
          
          <circle
            class="ng-star-inserted"
            cx="50%"
            cy="50%"
            r="45"
            style="animation-name: mat-progress-spinner-stroke-rotate-100; stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
          />
          
        </svg>
      </mat-spinner>
    </div>
    <mat-paginator
      class="mat-paginator"
    >
      <div
        class="mat-paginator-outer-container"
      >
        <div
          class="mat-paginator-container"
        >
          
          <div
            class="mat-paginator-page-size ng-star-inserted"
          >
            <div
              class="mat-paginator-page-size-label"
            >
              Items per page:
            </div>
            
            <mat-form-field
              class="mat-paginator-page-size-select mat-form-field ng-tns-c11-1 mat-primary mat-form-field-type-mat-select mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-star-inserted"
            >
              <div
                class="mat-form-field-wrapper"
              >
                <div
                  class="mat-form-field-flex"
                >
                  
                  
                  <div
                    class="mat-form-field-infix"
                  >
                    <mat-select
                      aria-disabled="false"
                      aria-invalid="false"
                      aria-label="Items per page:"
                      aria-multiselectable="false"
                      aria-required="false"
                      class="mat-select ng-tns-c12-2 mat-select-empty ng-star-inserted"
                      id="mat-select-1"
                      role="listbox"
                      tabindex="0"
                    >
                      <div
                        aria-hidden="true"
                        cdk-overlay-origin=""
                        class="mat-select-trigger"
                      >
                        <div
                          class="mat-select-value"
                        >
                          
                          <span
                            class="mat-select-placeholder ng-tns-c12-2 ng-star-inserted"
                          >
                            Â 
                          </span>
                          
                        </div>
                        <div
                          class="mat-select-arrow-wrapper"
                        >
                          <div
                            class="mat-select-arrow"
                          />
                        </div>
                      </div>
                      
                    </mat-select>
                    <span
                      class="mat-form-field-label-wrapper"
                    >
                      
                    </span>
                  </div>
                  
                </div>
                
                <div
                  class="mat-form-field-underline ng-tns-c11-1 ng-star-inserted"
                >
                  <span
                    class="mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-form-field-subscript-wrapper"
                >
                  
                  
                  <div
                    class="mat-form-field-hint-wrapper ng-tns-c11-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                    style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                  >
                    
                    <div
                      class="mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
            
          </div>
          <div
            class="mat-paginator-range-actions"
          >
            <div
              class="mat-paginator-range-label"
            >
              1 - 10 of 100
            </div>
            
            <button
              aria-describedby="cdk-describedby-message-2"
              aria-label="Previous page"
              cdk-describedby-host=""
              class="mat-paginator-navigation-previous mat-icon-button"
              disabled=""
              mat-icon-button=""
              type="button"
            >
              <span
                class="mat-button-wrapper"
              >
                <svg
                  class="mat-paginator-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </span>
              <div
                class="mat-button-ripple mat-ripple mat-button-ripple-round"
                matripple=""
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            <button
              aria-describedby="cdk-describedby-message-3"
              aria-label="Next page"
              cdk-describedby-host=""
              class="mat-paginator-navigation-next mat-icon-button"
              mat-icon-button=""
              type="button"
            >
              <span
                class="mat-button-wrapper"
              >
                <svg
                  class="mat-paginator-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </span>
              <div
                class="mat-button-ripple mat-ripple mat-button-ripple-round"
                matripple=""
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            
          </div>
        </div>
      </div>
    </mat-paginator>
  </div>
</app-coin-market>
`;
